services/web/server/src/simcore_service_webserver/exporter/utils.py

```diff
@@ -0,0 +1,64 @@
+import asyncio
+import tempfile
+
+from pathlib import Path
+from typing import Union, Optional
+
+import os
+from aiofiles import os as aiofiles_os
+
+from aiohttp.web import FileResponse
+from aiohttp.typedefs import LooseHeaders
+from aiohttp.abc import AbstractStreamWriter
+
+makedirs = aiofiles_os.wrap(os.makedirs)  # as in aiofiles.os.py module
+rename = aiofiles_os.wrap(os.rename)  # as in aiofiles.os.py module
+path_getsize = aiofiles_os.wrap(os.path.getsize)  # as in aiofiles.os.py module
+
+
+def _candidate_tmp_dir() -> Path:
+    # pylint: disable=protected-access
+    # let us all thank codeclimate for this beautiful piece of code
+    return Path("/") / f"tmp/{next(tempfile._get_candidate_names())}"
+
+
+async def get_empty_tmp_dir() -> str:
+    candidate = _candidate_tmp_dir()
+    while candidate.is_dir() or candidate.is_file() or candidate.is_symlink():
+        candidate = _candidate_tmp_dir()
+
+    await makedirs(candidate, exist_ok=True)
+
+    return str(candidate)
+
+
+async def remove_dir(directory: str) -> None:
+    await asyncio.create_subprocess_exec("rm", "-rf", directory)
+
+
+class CleanupFileResponse(FileResponse):  # pylint: disable=too-many-ancestors
+    """After the upload is completed it will schedule a callback"""
+
+    def __init__(
+        self,
+        temp_dir: str,
+        path: Union[str, Path],
+        chunk_size: int = 256 * 1024,
+        status: int = 200,
+        reason: Optional[str] = None,
+        headers: Optional[LooseHeaders] = None,
+    ) -> None:
+        super().__init__(
+            path=path,
+            chunk_size=chunk_size,
+            status=status,
+            reason=reason,
+            headers=headers,
+        )
+        self._temp_dir = temp_dir
+
+    async def prepare(self, request: "BaseRequest") -> Optional[AbstractStreamWriter]:
+        try:
+            return await super().prepare(request=request)
+        finally:
+            await asyncio.get_event_loop().create_task(remove_dir(self._temp_dir))
```

Comment:
To have a working version of this we need to push all these paths onto Redis, and make sure they are removed after let's say UPLOAD/DOWNLOAD timeout + 1 hour. This way the GC will remove them.
Dose this sound good to you?

Reply:
one question, when does this call to remove the directory happen? If I export multiple times. what happens to the storage?

Reply:
1. It might not be that straight forward, This call happens only after the request ends, it dose not matter what is the status of the response. The cleanup is always guaranteed to run.
2. Exporting multiple studies in parallel is fully supported, this is a summary of how the disk is handled :
    - Each time an export request is created a new temporary directory is generated `/tmp/SOME_TMP_DIR/ `.
    -  Inside this directory  a new directory is generated with the uuid of the project `/tmp/SOME_TMP_DIR/uuid/`. All contents from the project are written inside the `/tmp/SOME_TMP_DIR/uuid/` directory (the content comes from storage, the manifest and the projects table).
    - The `/tmp/SOME_TMP_DIR/uuid/` is zipped producing the archive to be downloaded in this path`/tmp/SOME_TMP_DIR/some_name#SHA256=SOME_HASH.osparc`
    - When the request finishes, for any reason (HTTO_OK, HTTP_ERROR, etc...) the `/tmp/SOME_TMP_DIR/` si removed from the disk.